src/metabase/query_processor/parameters.clj

```diff
@@ -0,0 +1,248 @@
+      (substitute-all-params query-dict rest)
+      (let [query (update-in query-dict [:native :query] (partial substitute-param param-name value))]
+        (substitute-all-params query rest)))
+    query-dict))
+
+(def ^:private ^:const outer-clause-pattern #"\[\[.*?\]\]")
+(def ^:private ^:const outer-clause-prefix-pattern #"^\[\[(.*?)\s.*\]\]$")
+(def ^:private ^:const incomplete-outer-clause-pattern #"\[\[.*?\{\{.*?\}\}.*?\]\]")
+(def ^:private ^:const inner-clause-pattern #"<(.*?)>")
+(def ^:private ^:const incomplete-inner-clause-pattern #"<.*?\{\{.*?\}\}.*?>")
+
+(defn- remove-incomplete-clauses
+  "Scans the native query body and removes any custom clauses which were not substituted with a parameter."
+  [query-dict]
+  (let [find-and-replace (fn [sql]
+                           (-> sql
+                               (s/replace incomplete-outer-clause-pattern "")
+                               (s/replace incomplete-inner-clause-pattern "")))]
+    (update-in query-dict [:native :query] find-and-replace)))
+
+(defn- format-outer-clause
+  "Formats an outer clause, finding all inner clauses and joining them via `AND` and then concatenating that onto the clause prefix.
+   e.g. \"[[WHERE <foo='bar'> <tou='can'>]]\" -> \"WHERE foo='bar' AND tou='can'\""
+  [clause]
+  (let [prefix (second (re-find outer-clause-prefix-pattern clause))]
+    ;; re-seq produces a vector for each match like [matched-form grouping1] and we only want grouping1.
+    (str prefix " " (s/join " AND " (map second (re-seq inner-clause-pattern clause))))))
+
+(defn- process-outer-clauses
+  "Takes a native query body and completely processes all included outer clauses present."
+  [query-dict]
+  (if-let [outer-clauses (re-seq outer-clause-pattern (get-in query-dict [:native :query]))]
+    (update-in query-dict [:native :query] (fn [q]
+                                             (loop [sql                   q
+                                                    [outer-clause & rest] outer-clauses]
+                                               (if outer-clause
+                                                 (recur (s/replace-first sql outer-clause (format-outer-clause outer-clause)) rest)
+                                                 sql))))
+    query-dict))
+
+(defn- process-inner-clauses
+  "Takes a native query body and completely processes all included inner clauses present.
+   e.g. \"<foo='bar'>\" -> \"foo='bar'\""
+  [query-dict]
+  (if-let [inner-clauses (re-seq inner-clause-pattern (get-in query-dict [:native :query]))]
+    (update-in query-dict [:native :query] (fn [q]
+                                             (loop [sql                      q
+                                                    [[orig stripped] & rest] inner-clauses]
+                                               (if orig
+                                                 (recur (s/replace-first sql orig stripped) rest)
```

Comment:
why can't you just replace all the inner clauses at once with `s/replace` instead of doing them one-at-a-time and recursing?